x-pack/plugins/security_solution/public/actions/show_top_n/show_top_n.tsx

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { CellActionExecutionContext } from '@kbn/ui-actions-plugin/public';
+import { createAction } from '@kbn/ui-actions-plugin/public';
+import { i18n } from '@kbn/i18n';
+import ReactDOM, { unmountComponentAtNode } from 'react-dom';
+import React from 'react';
+
+import type { History } from 'history';
+import { EuiThemeProvider } from '@kbn/kibana-react-plugin/common';
+import { Provider } from 'react-redux';
+import { Router } from 'react-router-dom';
+
+import { KibanaContextProvider, KibanaServices } from '../../common/lib/kibana';
+import { APP_NAME, DEFAULT_DARK_MODE } from '../../../common/constants';
+import type { SecurityAppStore } from '../../common/store';
+import { isInSecurityApp } from '../utils';
+import { TopNAction } from './show_top_n_component';
+
+const SHOW_TOP = (fieldName: string) =>
+  i18n.translate('xpack.securitySolution.actions.showTopTooltip', {
+    values: { fieldName },
+    defaultMessage: `Show top {fieldName}`,
+  });
+
+const ID = 'security_showTopN';
+const ICON = 'visBarVertical';
+const UNSUPPORTED_FIELD_TYPES = ['date', 'text'];
+
+export interface ShowTopNActionContext extends CellActionExecutionContext {
+  metadata?: {
+    scopeId?: string;
+  };
+}
+
+export const createShowTopNAction = ({
+  store,
+  history,
+  order,
+}: {
+  store: SecurityAppStore;
+  history: History;
+  order?: number;
+}) => {
+  let currentAppId: string | undefined;
+
+  const services = KibanaServices.get();
+
+  services.application.currentAppId$.subscribe((appId) => {
+    currentAppId = appId;
+  });
+
+  return createAction<ShowTopNActionContext>({
+    id: ID,
+    type: ID,
+    order,
+    getIconType: (): string => ICON,
+    getDisplayName: ({ field }) => SHOW_TOP(field.name),
+    getDisplayNameTooltip: ({ field }) => SHOW_TOP(field.name),
+    isCompatible: async ({ field }) =>
+      isInSecurityApp(currentAppId) &&
+      field.name != null &&
+      field.value != null &&
+      !UNSUPPORTED_FIELD_TYPES.includes(field.type),
+    execute: async (context) => {
+      const onClose = () => {
+        if (context.extraContentNodeRef.current !== null) {
+          unmountComponentAtNode(context.extraContentNodeRef.current);
+        }
+      };
+
```

Comment:
I am not proud of rendering all these providers here. But the only alternative would be to refactor the UiActions plugin interface to allow an action to export a component and ensure it doesn't lose the Application context.

Reply:
Is the limitation in the `createAction` function? I think we can have any extra property in the `Action` class. Couldn't we establish that if an action (CellAction) has a `component` or `render` function defined, we use it?

Reply:
Well, my previous comment is not 100% accurate, we can add public attributes to the Action implementations, but they won't be accessible in a generic way from the cellActions component. 
Actually, It is awesome you found a working solution for the topN popover, let's iterate over it a bit more. Maybe we can simplify it further, so we don't need to add these annoying providers 👍